float32 phi
float32 phi_dot